### PR TITLE
Add parameter passing support for xslt transformations

### PR DESCRIPTION
**Pull request by lankavitharana**
## Purpose
Add parameter passing support for XSLT transformations (similar to EI)

## Examples
```
xml input = xml `<abc></abc>`;
xml xltVal = xml `<xsl:stylesheet version="1.0" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">
<xsl:param name="param1"></xsl:param>
    <xsl:template match="/">
        <test>
                    <xsl:value-of select="$param1"/>
        </test>
    </xsl:template>
</xsl:stylesheet>`;
map<string|int> params = {"param1":"value"};
xml target = check xslt:transform(input, xltVal, params);
```

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [ ] Checked native-image compatibility

### CODE REVIEW

**Comment by lankavitharana** [comment]
Hi @daneshk, 

Can you please have a look at this? appreciate if if you can get this feature in

Thank you,
Rajith

**Comment by ThisaruGuruge** [comment]
Sorry about the delay @lankavitharana. We will check this and get back.

**Comment by daneshk** [comment]
Thanks for the fix @lankavitharana. @Bhashinee could you please check

File: ballerina/natives.bal [hunk 1]
```diff
@@ -23,8 +23,9 @@ import ballerina/jballerina.java;
 #
 # + input - An XML object, which needs to be transformed
 # + xsl - The XSL style sheet represented in an XML object
+# + params - Optional parameters passed into the xslt template
 # + return - The transformed result represented in an XML object or else an `error` if the given `XML` object
 #            cannot be transformed
-public isolated function transform(xml input, xml xsl) returns xml|error = @java:Method {
+public isolated function transform(xml input, xml xsl, map<string|int> params = {}) returns xml|error = @java:Method {
```
Comment (Bhashinee) [inline ballerina/natives.bal:29]: Instead of the `int` type, isn't it better to have `decimal` type here?

**Comment by lankavitharana** [comment]
Hi @ThisaruGuruge, were you able to have a look at the PR?

File: ballerina/tests/read-from-file.bal [hunk 2]
```diff
@@ -34,6 +34,45 @@ function testReadFromFile() {
...
 
+@test:Config {}
+function testReadFromFileWithParams() {
+    string xmlFilePath = "tests/resources/datafiles/cd_catalog.xml";
+    string xslFilePath = "tests/resources/datafiles/cd_catalog_with_params.xsl";
+    string strParam = "Music CD Collection With Artist";
+    decimal intParam = 5;
+    map<string|decimal> params = {"param1":strParam, "param2": intParam};
+    xml|error result = readFromFileWithParams(xmlFilePath, xslFilePath, params);
+    if (result is xml) {
```
Comment (Bhashinee) [inline ballerina/tests/read-from-file.bal:45]:
```suggestion
    if result is xml {
```

File: ballerina/tests/read-from-file.bal [hunk 3]
```diff
@@ -34,6 +34,45 @@ function testReadFromFile() {
...
+        xml|error expected = readXml("tests/resources/datafiles/read_from_file_with_params_result.xml");
+        if (expected is xml) {
```
Comment (Bhashinee) [inline ballerina/tests/read-from-file.bal:47]:
```suggestion
        if expected is xml {
```

File: ballerina/tests/read-from-file.bal [hunk 4]
```diff
@@ -34,6 +34,45 @@ function testReadFromFile() {
...
+            test:assertFail(expected.message());
+        }
+    } else {
+        test:assertFail(result.message());
+    }
+}
+
+function readFromFileWithParams(string xmlFilePath, string xslFilePath, map<string|decimal> params) returns xml|error {
+    var xmlValue = readXml(xmlFilePath);
+    if (xmlValue is xml) {
```
Comment (Bhashinee) [inline ballerina/tests/read-from-file.bal:59]:
```suggestion
    if xmlValue is xml {
```

File: ballerina/tests/read-from-file.bal [hunk 5]
```diff
@@ -34,6 +34,45 @@ function testReadFromFile() {
...
+        var xslValue = readXml(xslFilePath);
+        if (xslValue is xml) {
```
Comment (Bhashinee) [inline ballerina/tests/read-from-file.bal:61]:
```suggestion
        if xslValue is xml {
```

File: ballerina/tests/read-from-file.bal [hunk 6]
```diff
@@ -34,6 +34,45 @@ function testReadFromFile() {
...
+        xml|error expected = readXml("tests/resources/datafiles/read_from_file_with_params_result.xml");
+        if (expected is xml) {
+            test:assertEquals(result, expected);
+        } else {
+            test:assertFail(expected.message());
+        }
+    } else {
+        test:assertFail(result.message());
+    }
+}
```
Comment (Bhashinee) [inline ballerina/tests/read-from-file.bal:55]:
```suggestion
function testReadFromFileWithParams() returns error? {
    string xmlFilePath = "tests/resources/datafiles/cd_catalog.xml";
    string xslFilePath = "tests/resources/datafiles/cd_catalog_with_params.xsl";
    string strParam = "Music CD Collection With Artist";
    decimal intParam = 5;
    map<string|decimal> params = {"param1":strParam, "param2": intParam};
    xml result = check readFromFileWithParams(xmlFilePath, xslFilePath, params);
    xml expected = check readXml("tests/resources/datafiles/read_from_file_with_params_result.xml");
    test:assertEquals(result, expected);
}
```

File: ballerina/tests/read-from-file.bal [hunk 7]
```diff
@@ -34,6 +34,45 @@ function testReadFromFile() {
...
+            } else {
+                return result;
+            }
+        } else {
+            return xslValue;
+        }
+    } else {
+        return xmlValue;
+    }
+}
```
Comment (Bhashinee) [inline ballerina/tests/read-from-file.bal:74]:
```suggestion
function readFromFileWithParams(string xmlFilePath, string xslFilePath, map<string|decimal> params) returns xml|error {
    xml xmlValue = check readXml(xmlFilePath);
    xml xslValue = check readXml(xslFilePath);
    return transform(xmlValue, xslValue, params);
}
```

File: ballerina/natives.bal (hunk 1 above)
Comment (Bhashinee) [reply ballerina/natives.bal]: Updated the code to replace the `int` type with the `decimal` type.
[466c706](https://github.com/ballerina-platform/module-ballerina-xslt/pull/469/commits/466c706d9862d01056b7b081c6a5ab156a4ab739)

**Review by Bhashinee — APPROVED** [review]

**Comment by lankavitharana** [comment]
Hi @Bhashinee, @daneshk, 

I am not sure how this works? should I merge this? or are you guys going to do that?,

Thank you,
Rajith

**Pull request merged**